url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPSystemManager.java

```diff
@@ -0,0 +1,1344 @@
+                            + link.linkState);
+                    break;
+                }
+            }
+        }
+
+        if (state != newState) {
+            switch (newState) {
+            case CONNECTING:
+            case INCONSISTENT:
+                raiseCannotTakeCalls(newState);
+                break;
+
+            case MISSING_MASTER:
+                raiseCannotTakeCalls(newState);
+                missingMasterStart = System.currentTimeMillis();
+                break;
+
+            case MASTER_SLAVE:
+                preferredLink = master;
+                break;
+
+            case LOADSHARING:
+                break;
+
+            case STOPPED: // Clear alarms when stopped
+            case NO_LINKS: // Clear alarm as this may be intentional
+                alarms.clearAlarm(ALARM_CANNOT_TAKE_CALLS, this);
+                break;
+
+            default:
+                log.debug("Unexpected newState " + newState);
+                break;
+            }
+        }
+
+        if (newState == SystemState.MASTER_SLAVE
+                || newState == SystemState.LOADSHARING) {
+            // Check if at least one link is active
+            boolean foundActiveLink = false;
+            for (Link link : links) {
+                if (link.isUnitActive()) {
+                    foundActiveLink = true;
+                    break;
+                }
+            }
+
+            if (foundActiveLink) {
+                alarms.clearAlarm(ALARM_CANNOT_TAKE_CALLS, this);
+                gappingAlarmActive = false;
+            } else if (!gappingAlarmActive) {
+                raiseCannotTakeCalls(newState);
+                gappingAlarmActive = true;
+            }
+        }
+
+        return newState;
+    }
+
+    /**
+     * Raise the ocpCannotTakeCalls alarm.
+     *
+     * @param newState the state that the system manager is about to enter
+     */
+    private void raiseCannotTakeCalls(final SystemState newState) {
+        String[] params = new String[links.size() + 2];
+        params[0] = newState.toString();
+        params[1] = Integer.toString(links.size());
+        for (int i = 0; i < links.size(); i++) {
+            params[i + 2] = links.get(i).linkState.toString();
+        }
+
+        alarms.raiseAlarm(ALARM_CANNOT_TAKE_CALLS, this, params);
+    }
+
+    /**
+     * Recalculate the preferred unit. The preferred unit is chosen as follows:
+     * <p/>
+     * <b>Master/slave:</b> The preferred unit is the master unit.
+     * <p/>
+     * <b>Loadsharing:</b> If there is only one link, then it is chosen.
+     * Otherwise, if all units report the same unit as being preferred and it is
+     * a unit that this object is connected to then it is chosen. Otherwise
+     * there is no preferred unit.
+     *
+     * @param newState the new system state to use for recalculating.
+     */
+    private void recalcPreferredUnit(final SystemState newState) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.